### PR TITLE
New versioning process (with multiple virtual environments)

### PR DESCRIPTION
**Pull request by jgoizueta**
This reorganizes the code and provides a development/releasing/deploying process which relies on using the system's python-sicpy package and having separete virtual environments for development and for each release

### CODE REVIEW

**Comment by jgoizueta** [comment]
cc @rafatower can you review? this is ready to be merged, version number and .pyc problems solved.
After merging this we'll do a sample feature release.

File: README.md [hunk 1]
```diff
@@ -4,9 +4,146 @@ CartoDB Spatial Analysis extension for PostgreSQL.
...
+This branch and `master` are maintained by the *Release Manager*.
+The `master` branch is used to merge and tag releases to be
+deployed in production.
+
+In addition to these two permanent branches, temporal *topic*
+branches will be used for all modifications.
+
+## Development
+
+A topic branch should be created out of the `develop` branch
```
Comment (rafatower) [inline README.md]: I'd use MUST/SHOULD/SHALL etc. in the sense of RFC standard https://www.ietf.org/rfc/rfc2119.txt

File: README.md [hunk 2]
```diff
@@ -4,9 +4,146 @@ CartoDB Spatial Analysis extension for PostgreSQL.
...
 ## Requirements
 
-* pip
+* pip, virtualenv, PostgreSQL
+* python-scipy system package (see src/py/README.md)
+
+# Working Process
+
+We use the default `develop` branch as the basis for development.
+This branch and `master` are maintained by the *Release Manager*.
```
Comment (rafatower) [inline README.md]: `develop` is not the responsibility of the release manager, is it?

File: README.md [hunk 3]
```diff
@@ -4,9 +4,146 @@ CartoDB Spatial Analysis extension for PostgreSQL.
...
+
+Note: we keep the current development version install as 'dev' always;
+we update through the 'current' alias to allow changing the extension
+contents but not the version identifier. This will fail if the
+changes involve incompatible function changes such as a different
+return type; in that case the offending function (or the whole extension)
+should be dropped manually before the update.
+
+If the extension has not previously been installed in a database
+we can:
```
Comment (rafatower) [inline README.md]: you can

File: README.md [hunk 4]
```diff
@@ -4,9 +4,146 @@ CartoDB Spatial Analysis extension for PostgreSQL.
...
+
+* `CREATE EXTENSION crankshaft WITH VERSION 'dev';`
+
+Note: the development extension uses the development pyhton virtual
+environment automatically.
+
+Once the tests are succeeding a new Pull-Request can be created.
+CI-tests must be checked to be successful.
+
+Before proceeding to the release process peer code reviewing of the code is a must.
```
Comment (rafatower) [inline README.md]: Pull Request (PR) to `develop` branch. I'd be the most prescriptive when defining the development process

File: README.md [hunk 5]
```diff
@@ -4,9 +4,146 @@ CartoDB Spatial Analysis extension for PostgreSQL.
...
+
+## Release
+
+The release process of a new version of the extension
+shall by performed by the designated *Release Manager*.
```
Comment (rafatower) [inline README.md]: shall be

Comment (rafatower) [inline README.md]: the release part can probably be moved to a separate doc, but ok keeping it all together here FTM

File: README.md [hunk 6]
```diff
@@ -4,9 +4,146 @@ CartoDB Spatial Analysis extension for PostgreSQL.
...
+releases (updating the minor or patch version numbers).
+
+TODO: document the complex case of major releases.
+
+The next command must be executed to produce the main installation
+script for the new release, `release/cranckshaft--X.Y.Z.sql` and
+also to copy the python package to `release/python/X.Y.Z/crankshaft`.
+
+```
+make release
```
Comment (rafatower) [inline README.md]: some documentation about what each `make [target]` does might be good. Even in a `make help` target. Also warning about who should execute what.

File: src/pg/.gitignore [hunk 7]
```diff
@@ -0,0 +1,6 @@
+regression.diffs
+regression.out
+results/
+crankshaft--dev.sql
+crankshaft--dev--current.sql
+crankshaft--current--dev.sql
```
Comment (rafatower) [inline src/pg/.gitignore]: missing the *.pyc files

Comment (rafatower) [inline src/pg/.gitignore]: also I have my reservations about committing the virtualenv here. It shall only contain the crankshaft package, right?

File: src/pg/sql/01_version.sql [hunk 8]
```diff
@@ -0,0 +1,12 @@
+-- Version number of the extension release
+CREATE OR REPLACE FUNCTION cdb_crankshaft_version()
+RETURNS text AS $$
+  SELECT '@@VERSION@@'::text;
+$$ language 'sql' IMMUTABLE STRICT;
+
+-- Internal identifier of the installed extension instence
+-- e.g. 'dev' for current development version
+CREATE OR REPLACE FUNCTION cdb_crankshaft_internal_version()
```
Comment (rafatower) [inline src/pg/sql/01_version.sql]: isn't this private? `_cdb_crankshaft_internal_version`

I do believe so

File: src/pg/sql/01_version.sql [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+RETURNS text AS $$
+  SELECT installed_version FROM pg_available_extensions where name='crankshaft' and pg_available_extensions IS NOT NULL;
+$$ language 'sql' IMMUTABLE STRICT;
```
Comment (rafatower) [inline src/pg/sql/01_version.sql]: wouldn't the immutable be problematic when moving among versions?

Comment (rafatower) [inline src/pg/sql/01_version.sql]: I'd definitely remove the IMMUTABLE to avoid caching weird side effects

File: src/py/README.md [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+* At present we use Python version 2.7.3
+
+---
+
+We have two possible approaches being considered as to how manage
+the Python virtual environment: using a pure virtual enviroment
+or combine it with some system packages that include depencencies
+for the *hard-to-compile* packages (and pin them in somewhat old versions).
+
+### Alternative A: pure virtual environment
```
Comment (rafatower) [inline src/py/README.md]: I wouldn't discuss alternatives here. Let's just stick to the winner one.

File: src/py/crankshaft/setup.py [hunk 11]
```diff
@@ -10,7 +10,7 @@
 setup(
     name='crankshaft',
 
-    version='0.0.1',
+    version='0.0.0',
```
Comment (rafatower) [inline src/py/crankshaft/setup.py]: aren't the versions synced between sql and python

File: README.md [hunk 12]
```diff
@@ -4,9 +4,146 @@ CartoDB Spatial Analysis extension for PostgreSQL.
...
+* - *src/pg* contains the PostgreSQL extension source code
+* - *src/py* Python module source code
+* *release* reseleased versions
+* *env* base directory for Python virtual environments
 
 ## Requirements
 
-* pip
+* pip, virtualenv, PostgreSQL
+* python-scipy system package (see src/py/README.md)
```
Comment (rafatower) [inline README.md]: that's not actually required in the latest version, is it?

File: README.md (hunk 2 above)
Comment (jgoizueta) [inline README.md]: I think the developers should branch off develop and merge back to it, but al merges between develop and master and direct commits to develop should be handled by the RM (release manager)

File: src/pg/.gitignore (hunk 7 above)
Comment (jgoizueta) [inline src/pg/.gitignore]: the .pyc are now ignored in the top level .gitignore (may be in src/py as well as in release/python)

Comment (jgoizueta) [inline src/pg/.gitignore]: I also feel the virtualenv should not be handled in that makefile, but we need, e.g., the version information extracted from the .control to determine the name of the environment (and other things already computed here...). Considered doing it all at the top level but the particularities of pgxs complicate the matter... I'll give it another thought.

File: src/pg/sql/01_version.sql (hunk 8 above)
Comment (jgoizueta) [inline src/pg/sql/01_version.sql]: Yeah, right

File: src/pg/sql/01_version.sql (hunk 9 above)
Comment (jgoizueta) [inline src/pg/sql/01_version.sql]: OK. I'm not sure how IMMUTABLE is handled across extension versions/function redefinitions, but let's keep it safe...

File: src/py/README.md (hunk 10 above)
Comment (jgoizueta) [inline src/py/README.md]: Agreed, that's a a remnant to be removed

File: src/py/crankshaft/setup.py (hunk 11 above)
Comment (jgoizueta) [inline src/py/crankshaft/setup.py]: That 0.0.0 is replaced by the make file upon release. The package installed in the development environment keeps the 0.0.0, though (it is installed directly from the source code).

File: README.md (hunk 12 above)
Comment (jgoizueta) [inline README.md]: I think all that is required, we rely on the python-scipy (SciPy 0.9.0) to be able to install our PySAL 1.9.1 requirement

File: README.md (hunk 1 above)
Comment (jgoizueta) [inline README.md]: capitalization and all?

Comment (rafatower) [inline README.md]: no, but the unambiguous meaning: the process is that and that and that, not giving recommendations or maybes when "you **shall/must** follow these steps". The opposite is like giving them permission to break the process.

That's what I mean by being as prescriptive as possible.

File: README.md (hunk 2 above)
Comment (rafatower) [inline README.md]: agreed. Then please add that to the doc.

File: src/py/crankshaft/setup.py (hunk 11 above)
Comment (rafatower) [inline src/py/crankshaft/setup.py]: :+1:

File: README.md (hunk 12 above)
Comment (rafatower) [inline README.md]: you're right, sorry. I confused the libraries (once again). Just ignore my previous comment.

File: README.md [hunk 13]
```diff
@@ -4,9 +4,199 @@ CartoDB Spatial Analysis extension for PostgreSQL.
...
+
+# Working Process
+
+We use the default `develop` branch as the basis for development.
+This branch and `master` are maintained by the *Release Manager*.
+The `master` branch is used to merge and tag releases to be
+deployed in production.
+
+In addition to these two permanent branches, temporal *topic*
```
Comment (rafatower) [inline README.md]: I believe the correct adjective is "temporary"... anyway I'd rephrase this as: "A *topic branch* shall be created from `develop`  for any new feature or bugfix".

File: README.md [hunk 14]
```diff
@@ -4,9 +4,199 @@ CartoDB Spatial Analysis extension for PostgreSQL.
...
+branches must be used for all modifications.
+
+## Development
+
+For any modification of crankshaft, including adding new features,
+a topic branch must be created out of the `develop` branch
+and be used for the development process; see src/py/README.md
+for further details about the Pyhton package development.
```
Comment (rafatower) [inline README.md]: I do believe how branches are managed is independent of whether the changes lie in sql or python code, right?

File: README.md [hunk 15]
```diff
@@ -4,9 +4,199 @@ CartoDB Spatial Analysis extension for PostgreSQL.
...
+* `sudo make deploy` will install the current release X.Y.Z from the
+  `release/` files into PostgreSQL and a Python virtual environment
+  `envs/X.Y.Z`.
+  Intended for use by the release manager and deployment jobs.
+
+* `sudo make deploy RELEASE_VERSION=X.Y.Z` will install specified version
+  previously generated in `release/`
+  into PostgreSQL and a Python virtual environment `envs/X.Y.Z`.
+  Intended for use by the release manager and deployment jobs.
+```
```
Comment (rafatower) [inline README.md]: In order to keep things clean, I'd split this document into:
- `README.md`: scope of the extension and quickstart guide
- `CONTRIBUTING.md`: development process
- `RELEASE.md`: release process. Description of the RM role and responsiblities.

The scope is to be defined along with Research. But the point here is that for instance we'd need to split between functions generally available to all cartodb users and functions to be exposed through *services*.

I'd save this comment for the next iteration.

File: src/pg/sql/01_version.sql (hunk 8 above)
Comment (rafatower) [inline src/pg/sql/01_version.sql]: please fix it

File: README.md (hunk 14 above)
Comment (jgoizueta) [inline README.md]: yes

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -1,84 +1,90 @@
...
+environment automatically.
 
-## Testing
+Before proceeding to the release process peer code reviewing of the code is
+a must.
 
-Running just the Python tests:
+Once the feature or bugfix is completed, all the tests are passing
+and the code has been accepted by peer reviewing,
+the topic branch can be merged back into the `develop` branch and a
```
Comment (rafatower) [inline CONTRIBUTING.md]: Actually the order facilitated by github is PR first, then accept-merge the PR "a PR shall be created, reviewed by a peer and then merged to `develop` when CI tests pass"

File: README.md [hunk 17]
```diff
@@ -4,9 +4,69 @@ CartoDB Spatial Analysis extension for PostgreSQL.
...
+We distinguish two roles regarding the development cycle of crankshaft:
+
+* *developers* will implement new functionality and bugfixes into
+  the codebase and will request for new releases of the extension.
+* A *release manager* will attend these requests and will handle
+  the release process. The release process is sequential:
+  no concurrent releases will ever be in the works.
+
+We use the default `develop` branch as the basis for development.
+This branch and `master` are maintained by the *Release Manager*.
```
Comment (rafatower) [inline README.md]: after what we talked, just `master` is owned by the RM

File: README.md [hunk 18]
```diff
@@ -4,9 +4,69 @@ CartoDB Spatial Analysis extension for PostgreSQL.
...
+The `master` branch is used to merge and tag releases to be
+deployed in production.
```
Comment (rafatower) [inline README.md]: I'd avoid talking about deploys throughout this repo.

File: CONTRIBUTING.md (hunk 16 above)
Comment (jgoizueta) [inline CONTRIBUTING.md]: Right

File: README.md (hunk 18 above)
Comment (rafatower) [inline README.md]: never mind

**Comment by rafatower** [comment]
:+1: go, go! :smile:

File: README.md (hunk 17 above)
Comment (jgoizueta) [inline README.md]: Oops, missed that one
